택배 송장 제조기 [STEP Bonus] gogochang

gogochang opened the pull request:
@GREENOVER 
안녕하세요. 이번 PR에서는
객체 미용 체조 3,8원칙,
이벤트 할인 정책 추가
영수증 발송 추가에 대한 과정과 고민들을 작성했습니다.

## 고민했던 점
### 객체 미용 체조 3 원칙
- 가장 먼저 눈에 띄었던건 STEP1에서 새로운 타입으로 분리하면서 내부의 원시 타입 프로퍼티를 수정하지 않았던 코드였습니다. 이 원시 타입을 새로운 타입으로 포장하여 제가 원하는 값을 보장받기위해 수정했습니다.

- `deliveryCost`는 `Cost`, `receiverName`은 `Name` 타입 , `receiver`은 `Mobile`로 새로운 타입으로 포장하였고,
String 타입을  사용하는 `Name`와 `Mobile`은 정규표현식을 사용하여 값에 대한 검증을 하도록 하였고, Int를 사용하는 `Cost`는 0이상일 때만 값을 가지도록 작성했습니다. do, catch에 에러가 발생했을 때 어디서 무슨 에러인지 콘솔창에 확인하고자 새로운 `ParcelError` 열거형 타입도 추가해봤습니다.

- 작성하고 보니 중복된 코드가 보여 프로토콜로 공통된 코드를 제거할 수 있다고 생각했습니다.
`ValueWrapper`라는 새로운 프로토콜 타입을 생성하여 각 포장하는 타입들에게 프로토콜을 따르도록 추가하였습니다.
 값을 가져오는 메서드`getValue()`를 프로토콜 확장으로 메서드의 기본 구현하도록 작성하여 각 포장타입 내부의 중복된 메서드 코드를 제거할 수 있었습니다. 공통된 기능 외에는 각 포장 구조체 타입 내부에 추가하여 사용하도록 했습니다.

- `ValueWrapper`로 포장된 타입은 관리측면에 유리하도록 하나의 파일 내부에 작성하여 관리하도록 하였습니다.

```swift
protocol ValueWrapper {
    associatedtype valueType
    var value: valueType { get }
    func getValue() -> valueType
}

extension ValueWrapper {
    func getValue() -> valueType {
        return value
    }
}

struct Cost: ValueWrapper {
    let value: Double
    init(_ value: Double) throws {
        guard value > 0 else {
            throw ParcelError.cost
        }
        self.value = value
    }
}

struct Mobile: ValueWrapper {
    var value: String
    
    private let predicate = NSPredicate(
        format: "SELF MATCHES %@", "^01[0-1, 7][0-9]{7,8}$")
    
    init(_ value: String) throws {
        guard predicate.evaluate(with: value) else {
            throw ParcelError.mobile
        }
        self.value = value
    }
}

struct Name: ValueWrapper {
    var value: String
    
    private let predicate = NSPredicate(
        format: "SELF MATCHES %@", "^[가-힣a-zA-Z]{1,10}$")
    
    init(_ value: String) throws {
        guard predicate.evaluate(with: value) else {
            throw ParcelError.name
        }
        self.value = value
    }
}

```

### 객체 미용 체조 8 원칙
- 야곰 iOS 리팩터링 코스에 참여하지 않았다면, 그냥 넘어갈 수 있었던 코드였을지도 모르겠습니다. STEP2의 객체 미용 체조 2원칙 if-else, switch를 제거하면서 전략패턴을 사용하면서 외부에서 DiscountStrategy 배열을 사용한 부분에 냄새를 맡을 수 있었습니다. 8원칙 일급 콜렉션을 사용하여 하나의 타입으로 포장하여 관리할 수 있도록 하였습니다.

- `DiscountStrategy`배열을 프로퍼티로 가지는 `DiscountStrategyManager` 클래스 타입을 추가했습니다. 필요할만한 요구사항들을 메서드로 생성하여 외부에서 프로퍼티를 직접 접근하지 못하도록 포장하여 관리에 용이할 수 있도록 하였습니다.
 
```swift
// 객체 미용 체조, 8원칙 일급 콜렉션 사용
class DiscountStrategyManager {
    private var items: [DiscountStrategy]
    
    init(items: [DiscountStrategy]) {
        self.items = items
    }
    
    // 할인 정책들을 반환합니다.
    func getStrategies() -> [DiscountStrategy] {
        return items
    }
    
    // 특정 할인 정책이 아이템에 존재하는지 확인합니다.
    func isContainsStrategy<T: DiscountStrategy>(_ strategyType: T.Type) -> Bool {
        return items.contains { $0 is T }
    }
    
    // 할인 정책을 삭제합니다.
    func removeStrategy<T: DiscountStrategy>(_ strategyType: T.Type) {
        guard let index = items.firstIndex(where: {
            $0 is T }) else {
            return
        }
        
        items.remove(at: index)
    }
    
    // 할인 정책을 추가합니다.
    func addStrategy(_ strategy: DiscountStrategy) {
        guard (items.contains {type(of: $0) == type(of: strategy)}) else { return }
        
        items.append(strategy)
    }
    
    // 할인 정책에 대한 요구사항에 따라...
}

```
### 이벤트 할인 정책 추가
- 사실 생각보다 할게 없었습니다. `EventDiscountStrategy` 타입을 생성하고, `discountSegmented`, `discountStrategyManager`에 추가했을 뿐.. 이번 Week1에서 다루는 내용을 잘 학습한다면 이렇게 활용하고, 이런 장점이 있다 라는 의미를 전달하고 있구나 라고 생각이 들었습니다.
```swift
struct EventDiscountStrategy: DiscountStrategy {
    var discountRate: Double
}

let discountStrategyManager = DiscountStrategyManager(items: [
            NoDiscountStrategy(discountRate: 1),
            VIPDiscountStrategy(discountRate: 0.8),
            CouponDiscountStrategy(discountRate: 0.5),
            EventDiscountStrategy(discountRate: 0.9)
])
```
### 영수증 발송 기능 추가
- 영수증 방식을 선택하는 버튼을 추가했고, 영수증과 관련된 책임하나만 가질 수 있도록 `Receipt`열거형 타입 추가하여 `SRP`원칙과 내부에 프로퍼티 외부로 전달할 수 있는 메서드를 추가, 영수증 전송 메서드를 가지는 `ReceipProcessor`의 상위 계층 `ReceiptManageable`를 소유하도록 하여 `DIP`원칙을 따르도록 작성했습니다.

- 이전 STEP2에서 protocol을 사용하여 `OCP`원칙을 따르는 `DiscountStrategy` 구현했었습니다. 이번 STEP에서는 enum을 사용한 `OCP`를 작성했습니다.
영수증 발송 방법은 일반적으로 메일,문자,직접수령,우편 에서 크게 추가될 가능성이 없다고 생각했습니다. 
(더 있을 수 있겠지만 할인 방법은 요구사항에 따라 가지각색인 반면에 영수증 발송하는 방법은 상대적으로 제한되어있다고 생각했습니다.) 그린님과 이야기했던 "적절하게 `OCP`를 작성하는것", 이 상황에서는 타입의 추가 가능성이 적은 `enum`을 사용한 `OCP`가 적절하다고 생각했습니다! 

---

## 해결이 되지 않은 점
### 폴더구조
- 좀 더 고민을 해봐야할것같습니다. UIView를`View`, UIViewController를 `Controller`, 데이터를 처리하는 부분을 `Models`로 그룹화시킨 후 `Models` 그룹 내부에 각 주제별로 그룹화를 시켰는데, 명확한 관심사로 구분되어있지 않은느낌입니다. 네이밍도 아직 애매하구요.. 좀 더 고민해서 커밋해보겠습니다..

---

## 조언을 얻고 싶은 부분
### Class vs Struct
- 위에서 객체 미용 체조 8 원칙를 적용하면서 `DiscountStrategyManager`를 클래스 타입으로 생성했습니다.
 `DiscountStrategy`는 할인방식을 가지는 타입이기 때문에 이 할인 타입에 따라 다른 기능이나 페이지에 공유하거나 참조해야할 가능성이 있을것 같습니다. 일급콜렉션은 내부 프로퍼티를 관리한다는 목적도 있기 때문에 기본적으로 내부 프로퍼티 불변성을 가지는 구조체타입은 적절하지 않다고 생각했습니다. 이러한 이유로 클래스타입으로 `DiscountStrategyManager`를 생성했습니다. 

- 기초적인 질문일수 있겠지만 클래스와 구조체를 실제 코드에 어떤 상황에 적용해야할지 고민해보니 아직은 생각보다 바로 떠오르지 않는것같습니다. 제 생각이 맞는지 또 실제로 어떤 기준으로 클래스와 구조체를 선택하시는지 그린님의 조언을 얻고싶습니다.

gogochang committed:
info.plist 경로 수정

gogochang committed:
[STEP Bonus] 3원칙 원시값과 포장
	- ValueWrapper 프로토콜을 사용하여 원시타입의 포장과 중복코드 제거

gogochang committed:
[STEP Bonus] ParcelInfo 접근제한 설정

gogochang committed:
[STEP Bonus] 객체 미용 체조 8원칙
	- DiscountStrategy의 배열을 입급 콜렉션 사용

gogochang committed:
[STEP Bonus] 이벤트 할인 정책 추가

gogochang committed:
[STEP Bonus] 영수증 추가
	- enum을 사용한 OCP

gogochang committed:
오탈자 수정

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/ParcelInfo/DiscountStrategyManager.swift:
```diff
@@ -0,0 +1,46 @@
+import Foundation
+
+// 객체 미용 체조, 8원칙 일급 콜렉션 사용
+class DiscountStrategyManager {
```
해당 클래스가 더 이상 상속될 경우가 없다고 판단되면 final을 붙여보는건 어떨까요~? 🙋🏻
성능 향상에 더 도움을 줄 수 있기에 고려해보셔도 좋을것 같습니다!

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/ParcelInfo/DiscountStrategyManager.swift:
```diff
@@ -0,0 +1,46 @@
+    }
+    
+    // 특정 할인 정책이 아이템에 존재하는지 확인합니다.
+    func isContainsStrategy<T: DiscountStrategy>(_ strategyType: T.Type) -> Bool {
+        return items.contains { $0 is T }
+    }
```
현재 중복된 할인 전략이 추가되는걸 방지하기 위한 코드로 보이는데, 사용부에서는 사실 초기화로 먼저 items을 넣어주고 있어서 초기화 시에 중복된 할인 전략이 들어오는것은 체크하지 못할것 같은데 어떻게 생각하시나요~? 🙋🏻

gogochang replied:
객체 미용 체조 8원칙에 일급 콜렉션으로 포장하기 위한 클래스타입으로 생성했었습니다. 단순히 일급 콜렉션으로 포장한다는것에 집중을 하다보니 제가 깊게 생각하지 못한부분인것 같습니다. 그린님의 의견에 동의합니다.

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/ParcelInfo/DiscountStrategyManager.swift:
```diff
@@ -0,0 +1,46 @@
+    // 할인 정책을 삭제합니다.
+    func removeStrategy<T: DiscountStrategy>(_ strategyType: T.Type) {
+        guard let index = items.firstIndex(where: {
+            $0 is T }) else {
```
is로 타입 검사를 하는것과 type(of:) 메서드를 통해 타입 검사를 하는것 어떤것에 차이가 있으며, 좀 더 정밀하게 의도하신것처럼 타입 검사를 하기 위해서는 어떤것이 적절할까요..?!

gogochang replied:
@GREENOVER 
* is로 타입 검사
    * is로 타입 검사할 때는 상속 받은 타입일 경우에도 True을 반환합니다. 
    * 상속받은 클래스의 타입도 True로 반환합니다.

* type(of:) 로 타입 검사 
    * 타입의 메타타입을 반환합니다.

코드를 이렇게 작성한 이유는 한인율과 상관없이 타입이 오직 하나만 존재해야합니다.
그러려면 전달인자로 인스턴스가아닌 할인 정책이라고 추가한 타입(DiscountStrategy 프로토콜을 준수하는 타입)의 타입(메타타입)을 받아야 할인율과 상관없이 동일한 타입이 존재하는지 확인할 수 있습니다.

반대로 `addStrategy`메서드와 같이 전달받은 인스턴스가 존재하는지 확인하고자 합니다.
인스턴스로 부터 메타타입을 가져와서 사용할 때 `type(of)`메서드가 좀 더 좋을것같습니다.

확인하고자 하는 타입이 서브클래스라면 `is` 타입캐스팅을  사용하는것이 좋지 않을까 생각이듭니다.
또 코드가 더 간결해지고 가독성 면에 유리한장점으로 사용할 수있다면 `is`를 기본적으로 사용할것같습니다.

많이 헷갈립니다. 좀 더 찾아보고 보완,정리가 필요할것같습니다..

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/ParcelInfo/ParcelInfo.swift:
```diff
@@ -8,25 +8,43 @@
+    private let cost: ParcelCost
+    private let receipt: Receipt
+    
+    init(address: String, receiver: ReceiverInfo, cost: ParcelCost, receipt: Receipt) {
```
LGTM 👍👍👍

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/ParcelInfo/Receipt.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    // 영수증 발송 방식을 반환합니다.
+    func getSendMethod() -> String {
+        return method
+    }
```
꼭 필요한가요!?
이미 연산 프로퍼티이기에 사실 외부에서 변경을 해줄 수 없는 부분이라 보이는데 한번 더 메서드로 만들어서 분리하는것이 어떻게보면 오버엔지니어링이지 않을까 싶습니다 🤔

gogochang replied:
아 맞습니다!..  특별한 이유는 없었습니다. 제가 연산프로퍼티의 기본특성을 놓친것같습니다.

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/Wrapper/ValueWrapper.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+struct Cost: ValueWrapper {
```
Cost, Mobile, Name을 별도 파일로 분리해보면 어떨까요!?
사실 해당 파일 네이밍만 보고 어떤 역할을 하는지 어떤것들을 담는지 유추하는것이 좋을텐데 현재로썬 유추하기 힘들어보여서요ㅠ
나중에 파일이 수백개가 되고 추후 유지보수가 필요할때 파일들로 명확하고 빠르게 서칭하기가 힘들지 않을까 싶습니다..!

gogochang replied:
ValueWrapper그룹 내에 ValueWrapper의 파일과 또 하위 Value 그룹으로 포장되어진 타입들이 무슨 데이터인지 좀 더 명확히 알 수있도록 이름을 변경했습니다.

GREENOVER commented on ParcelInvoiceMaker/ParcelInvoiceMaker/Models/Wrapper/ValueWrapper.swift:
```diff
@@ -0,0 +1,57 @@
+protocol ValueWrapper {
+    associatedtype valueType
+    var value: valueType { get }
+    func getValue() -> valueType
```
extension에서 작성되어 있기에 해당 코드는 여기서 제거되어도 되지 않을까요~?
아니면 프로토콜을 타고 들어갔을때 해당 요구 명세를 더 명확하게 보기 위함이신지 다른 의도하신것이 있으실까요..!?

gogochang replied:
네 맞습니다. 프로토콜이 청사진 이기때문에 `getValue()`라는 메서드를 한번 더 작성해주었습니다. 말씀대로 바로 extension으로 작성을 한다면 제가 의도했던 것과 크게 다를게 없을것같습니다.

GREENOVER reviewed (approved):
@gogochang 
수고 많으셨습니다!
일주일이라는 짧은 기간동안 정말 빠르고 꼼꼼하게 리팩토링을 해주셨네요 👍
보너스 스텝의 요구사항은 만족하여 요구사항에 어긋난 부분보다는 한번 더 생각해볼 포인트들에 리뷰를 남겨보았습니다ㅎㅎ
***
### 궁금하신 부분에 대한 저의 의견
- Class vs Struct
  - 우선 저의 경우에서 둘을 사용하는 기준은 아마 뻔하고 아실 수 있겠지만, 클래스 사용을 택하는 상황들은 상속이 필요하거나 여러 곳에서 같은 인스턴스 참조를 요한다던가 등이 기본적일 수 있겠습니다.
  더 나아간다면 구조체는 사실 deinit이 되지 않기에 혹시 deinit을 사용해야하는 환경이다라면 클래스를 사용할 수 있겠고 타입 캐스팅이 필요하다던가 할때도 클래스를 택할 수 있을것 같아요.
  이런 케이스들이 걸리지 않는다면 클래스를 사용할 이유가 없으니 구조체를 사용하겠구요..!
  정말 정답은 없지만, 꼭 이런 경우가 아니더라도 구조체를 사용하는것은 아니에요.
  하나의 예시로, 구조체 내 프로퍼티가 참조 타입이면 RC가 발생하면서 갯수에 따라 오버헤드가 날 경우도 있습니다.
  (보통 한개 정도는 성능에 지장이 없다고 하는데..)
  그렇기에 구조체에서 만약 여러개의 프로퍼티를 레퍼런스로 사용한다면 클래스보다 오히려 성능이 안좋아질 수 있음으로 이럴 경우에는 구조체보단 클래스를 택하기도 합니다.
  이런것들을 볼때 딱 부러지지 않는 명확한 기준보다는 전체적인 흐름과 내 프로젝트에 맞는 상황을 보고 조금 유동적으로 선택하게 되는것 같습니다 😃
  지금 고고창이 근거를 가지고 선택하신 이유처럼요..!ㅎㅎ
***
다시 한번 수고 많으셨습니다!!
우선 어프로브 해두겠습니다~
언제든 편하게 머지 요청해주세요 😃

gogochang committed:
[STEP Bonus] 추가수정
	- 상속하지 않는 Class는 final 키워드 추가
	- DiscountStrategyManager의 초기화부분에서 중복 타입 검증
	- 연산 프로퍼티의 오버엔지니어링 제거
	- ValueWrapper 네이밍 명확하게 수정

GREENOVER commented:
@gogochang 
수고 많으셨습니다!!
